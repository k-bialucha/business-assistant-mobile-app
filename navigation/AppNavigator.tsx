import React from 'react';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { createStackNavigator } from '@react-navigation/stack';

import DashboardScreen from '../screens/DashboardScreen';
import SalesScreen from '../screens/SalesScreen';
import CostsScreen from '../screens/CostsScreen';
import CostsEntryScreen from '../screens/CostsEntryScreen';
import SalesEntryScreen from '../screens/SalesEntryScreen';

const DashboardStackNavigator = createStackNavigator();

const DashboardNavigator = () => {
  return (
    <DashboardStackNavigator.Navigator>
      <DashboardStackNavigator.Screen
        name="Dashboard"
        component={DashboardScreen}
      />
    </DashboardStackNavigator.Navigator>
  );
};

const SalesStackNavigator = createStackNavigator();

const SalesNavigator = () => {
  return (
    <SalesStackNavigator.Navigator>
      <SalesStackNavigator.Screen name="SalesList" component={SalesScreen} />
      <SalesStackNavigator.Screen
        name="SalesEntry"
        component={SalesEntryScreen}
      />
    </SalesStackNavigator.Navigator>
  );
};

const CostsStackNavigator = createStackNavigator();

const CostsNavigator = () => {
  return (
    <CostsStackNavigator.Navigator>
      <CostsStackNavigator.Screen name="CostsList" component={CostsScreen} />
      <CostsStackNavigator.Screen
        name="CostsEntry"
        component={CostsEntryScreen}
      />
    </CostsStackNavigator.Navigator>
  );
};

const AppTabNavigator = createBottomTabNavigator();

const AppNavigator = () => {
  return (
    <AppTabNavigator.Navigator>
      <AppTabNavigator.Screen name="Dashboard" component={DashboardNavigator} />
      <AppTabNavigator.Screen name="Sales" component={SalesNavigator} />
      <AppTabNavigator.Screen name="Costs" component={CostsNavigator} />
    </AppTabNavigator.Navigator>
  );
};

export default AppNavigator;
