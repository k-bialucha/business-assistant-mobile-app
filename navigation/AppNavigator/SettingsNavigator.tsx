import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import AuthScreen from '../../screens/AuthScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  Settings: undefined;
};

const SettingsStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Settings'>;

const SettingsNavigator: React.FC<Props> = () => {
  return (
    <SettingsStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <SettingsStackNavigator.Screen name="Settings" component={AuthScreen} />
    </SettingsStackNavigator.Navigator>
  );
};

export default SettingsNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Settings'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
