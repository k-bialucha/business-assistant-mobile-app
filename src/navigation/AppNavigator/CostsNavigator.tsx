import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import CostsEntryScreen from '~/screens/CostsEntryScreen';
import CostsListScreen, {
  CostsListScreenNavOptions,
} from '~/screens/CostsListScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  CostsList: undefined;
  CostsEntry: undefined;
};

const CostsStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Costs'>;

const CostsNavigator: React.FC<Props> = () => {
  return (
    <CostsStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <CostsStackNavigator.Screen
        name="CostsList"
        component={CostsListScreen}
        options={CostsListScreenNavOptions}
      />
      <CostsStackNavigator.Screen
        name="CostsEntry"
        component={CostsEntryScreen}
      />
    </CostsStackNavigator.Navigator>
  );
};

export default CostsNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Costs'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
