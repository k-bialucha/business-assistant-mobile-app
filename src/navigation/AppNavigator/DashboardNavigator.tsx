import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import DashboardScreen from '~/screens/DashboardScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  Dashboard: undefined;
};

const DashboardStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Dashboard'>;

const DashboardNavigator: React.FC<Props> = () => {
  return (
    <DashboardStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <DashboardStackNavigator.Screen
        name="Dashboard"
        component={DashboardScreen}
      />
    </DashboardStackNavigator.Navigator>
  );
};

export default DashboardNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Dashboard'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
