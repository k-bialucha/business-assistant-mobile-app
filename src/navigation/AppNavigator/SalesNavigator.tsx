import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import SalesEntryScreen from '~/screens/SalesEntryScreen';
import SalesScreen, { SalesScreenNavOptions } from '~/screens/SalesScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  SalesList: undefined;
  SalesEntry: undefined;
};

const SalesStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Sales'>;

const SalesNavigator: React.FC<Props> = () => {
  return (
    <SalesStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <SalesStackNavigator.Screen
        name="SalesList"
        component={SalesScreen}
        options={SalesScreenNavOptions}
      />
      <SalesStackNavigator.Screen
        name="SalesEntry"
        component={SalesEntryScreen}
      />
    </SalesStackNavigator.Navigator>
  );
};

export default SalesNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Sales'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
