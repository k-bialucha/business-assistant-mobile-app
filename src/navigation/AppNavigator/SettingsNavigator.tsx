import React from 'react';

import { CompositeNavigationProp, RouteProp } from '@react-navigation/native';
import {
  createStackNavigator,
  StackNavigationProp,
} from '@react-navigation/stack';

import SettingsScreen from '~/screens/SettingsScreen';
import UserSettingsScreen from '~/screens/UserSettingsScreen';
import { UserSettingsScreenNavOptions } from '~/screens/UserSettingsScreen/UserSettingsScreen';

import defaultScreenOptions from './defaultScreenOptions';
import {
  AppNavigatorNavigationProp,
  NavigationData as AppNavigatorNavigationData,
} from './index';

type ParamList = {
  Settings: undefined;
  UserSettings: undefined;
};

const SettingsStackNavigator = createStackNavigator<ParamList>();

type Props = AppNavigatorNavigationData<'Settings'>;

const SettingsNavigator: React.FC<Props> = () => {
  return (
    <SettingsStackNavigator.Navigator screenOptions={defaultScreenOptions}>
      <SettingsStackNavigator.Screen
        name="Settings"
        component={SettingsScreen}
      />
      <SettingsStackNavigator.Screen
        name="UserSettings"
        component={UserSettingsScreen}
        options={UserSettingsScreenNavOptions}
      />
    </SettingsStackNavigator.Navigator>
  );
};

export default SettingsNavigator;

// combine navigation prop for nested navigators
// (AppNavigator data + this StackNavigator data)
export type NavigationPropCombined<
  RouteName extends keyof ParamList
> = CompositeNavigationProp<
  AppNavigatorNavigationProp<'Settings'>,
  StackNavigationProp<ParamList, RouteName>
>;

// export type so screens can get typing
export type NavigationData<RouteName extends keyof ParamList> = {
  navigation: NavigationPropCombined<RouteName>;
  route: RouteProp<ParamList, RouteName>;
};
