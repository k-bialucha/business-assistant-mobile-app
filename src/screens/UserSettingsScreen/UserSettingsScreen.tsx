import React, { useEffect } from 'react';

import { StackNavigationOptions } from '@react-navigation/stack';
import { Trans, useTranslation } from 'react-i18next';

import PlainText from '~/components/UI/PlainText';
import { NavigationData } from '~/navigation/AppNavigator/SettingsNavigator';

import { StyledView } from './UserSettingsScreen.styled';

type Props = NavigationData<'UserSettings'>;

const UserSettingsScreen: React.FC<Props> = ({ navigation }) => {
  const { t } = useTranslation();

  useEffect(() => {
    // passing already translated title it could be common way to set nav options in entire app
    navigation.setParams({ headerTitle: t('Account Settings') });
  }, [t, navigation]);

  return (
    <StyledView>
      <PlainText theme="dark">User Settings Screen</PlainText>
    </StyledView>
  );
};

export const UserSettingsScreenNavOptions: StackNavigationOptions = navData => {
  // 1 - to avoid passing prop from component, but problem with headerTitle type, it expects string
  // headerTitle: <Trans i18nKey="Account Settings" />,

  // 2 set passed title, navData need type
  return { title: navData.route.params?.headerTitle };
};

export default UserSettingsScreen;
