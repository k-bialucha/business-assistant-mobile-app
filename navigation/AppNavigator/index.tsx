import React from 'react';
import { Platform } from 'react-native';

import { Ionicons } from '@expo/vector-icons';
import { createBottomTabNavigator } from '@react-navigation/bottom-tabs';
import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';

import Colors from '../../theme/Colors';

import CostsNavigator from './CostsNavigator';
import DashboardNavigator from './DashboardNavigator';
import SalesNavigator from './SalesNavigator';
import SettingsNavigator from './SettingsNavigator';

export type RootStackParamList = {
  Dashboard: undefined;
  Sales: undefined;
  Costs: undefined;
  Settings: undefined;
};

export type ScreenProps<RouteName extends keyof RootStackParamList> = {
  navigation: StackNavigationProp<RootStackParamList, RouteName>;
  route: RouteProp<RootStackParamList, RouteName>;
};

const AppTabNavigator = createBottomTabNavigator<RootStackParamList>();

const AppNavigator: React.FC<{}> = () => {
  return (
    <AppTabNavigator.Navigator
      initialRouteName="Dashboard"
      screenOptions={({ route }) => ({
        tabBarIcon: ({ color, size }) => {
          const icons = {
            Dashboard: Platform.OS === 'android' ? 'md-stats' : 'ios-stats',
            Sales:
              Platform.OS === 'android' ? 'md-trending-up' : 'ios-trending-up',
            Costs:
              Platform.OS === 'android'
                ? 'md-trending-down'
                : 'ios-trending-down',
            Settings:
              Platform.OS === 'android' ? 'md-settings' : 'ios-settings',
          };

          return (
            <Ionicons name={icons[route.name]} size={size} color={color} />
          );
        },
      })}
      tabBarOptions={{
        showLabel: false,
        activeTintColor: Colors.navyBlue,
        inactiveTintColor: Colors.silver,
      }}
    >
      <AppTabNavigator.Screen name="Sales" component={SalesNavigator} />
      <AppTabNavigator.Screen name="Dashboard" component={DashboardNavigator} />
      <AppTabNavigator.Screen name="Costs" component={CostsNavigator} />
      <AppTabNavigator.Screen name="Settings" component={SettingsNavigator} />
    </AppTabNavigator.Navigator>
  );
};

export default AppNavigator;
