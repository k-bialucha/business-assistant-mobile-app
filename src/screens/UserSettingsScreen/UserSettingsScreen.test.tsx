import React from 'react';

import { shallow } from 'enzyme';

import { NavigationData } from '~/navigation/AppNavigator/SettingsNavigator';

import UserSettingsScreen from './UserSettingsScreen';

jest.mock('react-redux', () => {
  return {
    __esModule: true,
    useDispatch: jest.fn().mockImplementation(() => jest.fn()),
  };
});

type Props = NavigationData<'UserSettings'>;

const fakeProps: Props = {
  route: { key: '1234', name: 'UserSettings' },
  // @ts-ignore
  navigation: {},
};

describe('<UserSettingsScreen />', () => {
  it('matches the snapshot', () => {
    const tree = shallow(<UserSettingsScreen {...fakeProps} />);

    expect(tree).toMatchSnapshot();
  });
});
