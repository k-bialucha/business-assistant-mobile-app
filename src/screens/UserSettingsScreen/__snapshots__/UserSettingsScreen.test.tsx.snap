// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UserSettingsScreen /> matches the snapshot 1`] = `
<Styled(View)>
  <PlainText
    theme="dark"
  >
    User Settings Screen
  </PlainText>
</Styled(View)>
`;
